### ISSUE TITLE
Nullable Reference Types: Child classes should be responsible for initializing fields not initialized by abstract parents

### ISSUE THREAD

**Issue by PathogenDavid** [dotnet/roslyn#27015]
**Version Used**: 05/14/18 Nullable Reference Types Preview (`csc` reports `2.8.0.62830 (e595ee27)`) with Visual Studio 15.7.1

**Demonstration Code**:

```csharp
public abstract class AbstractBaseClass
{
    public string Property0 { get; protected set; }
    public string Property1 { get; protected set; }
    public string Property2 { get; private set; }
    public string Property3 { get; protected set; }

    // warning CS8618: Non-nullable property 'Property0' is uninitialized.
    // warning CS8618: Non-nullable property 'Property1' is uninitialized.
    // warning CS8618: Non-nullable property 'Property2' is uninitialized.
    public AbstractBaseClass()
        => Property3 = "3";
}

public class ChildClass : AbstractBaseClass
{
    public string Property4 { get; private set; }

    // warning CS8618: Non-nullable property 'Property4' is uninitialized.
    public ChildClass()
        => Property0 = "0";
}
```

**Expected Behavior**:

The constructor for `AbstractBaseClass` has a warning for `Property2` and no others.

The constructor for `ChildClass` has a warning for `Property1` and `Property4`.

**Actual Behavior**:

(The warnings received are as displayed in the comments in the example code.)

The constructor for `AbstractBaseClass` has a warning for `Property0`, `Property1`, and `Property2` despite `ChildClass` being capable of initializing `Property0` and `Property1`.

The constructor for `ChildClass` only receives a warning for `Property4`

----

This is potentially non-trivial since the abstract class could lie within another assembly. I feel like the easiest solution would be to add an attribute `[DefinitelyInitialized]` to fields/properties that are appropriately initialized by the abstract constructors. However, I don't know how this would handle multiple constructors since you could initialize different subsets of properties in different constructors.

Some people might consider relying on child implementations to perform initialization to be evil, but I'd consider it to be in the same realm as the problem described [on the wiki page](https://github.com/dotnet/csharplang/wiki/Nullable-Reference-Types-Preview#warnings-for-initialized-fields) with indirect initialization.

In our use-cases, creating the object needed by the base class is too complicated* to do before the base constructor is called, so we rely on child implementations assigning the property themselves.

(*In particular, we want to do some argument validation before constructing the object because it is a native resource and we don't want to leak its handle if we throw an exception after associating it with an `IDisposable`.)

**Comment by Joe4evr** [comment]
> I feel like the easiest solution would be to add an attribute `[DefinitelyInitialized]` to fields/properties that are appropriately initialized

I think it'd be nice if using the ~~dammit~~ null-forgiving operator would be a short-cut to exempt checking assignment of those properties.
```cs
public string! Property0 { get; protected set; }
public string! Property1 { get; protected set; }
public string  Property2 { get; private set; }
public string! Property3 { get; protected set; }
```

**Comment by PathogenDavid** [comment]
I'm not sure if special syntax for that situation would be necessary (since you can just initialize them with the damnit operator IE: `Property0 = null!;`, which is what I've been doing for now.) However, I can see it being nice when you have a lot of properties/fields that you don't want to initialize for some reason. (Although how often would that happen?)

------

>I feel like the easiest solution would be to add an attribute `[DefinitelyInitialized]` to fields/properties that are appropriately initialized by the abstract constructors.

For clarity++: I was saying that Roslyn should add these attributes, not the user. That way external consumers know which field they need to initialize in their base classes.

Also after thinking about it more, it probably makes more sense to put a `[NotInitializedInBase]` attribute instead since it's the exception and `[DefinitelyInitialized]` is the rule.

**Comment by bkoelman** [comment]
Sounds like an anti-pattern to require derived classes to initialize their base class properly. Types should be responsible for their own internal state.

**Comment by Joe4evr** [comment]
> Sounds like an anti-pattern to require derived classes to initialize their base class properly.

True, but here's other use-cases that can benefit from having this extra annotation as well. For example, when working with an API that transiently `new`'s up objects per request (just like MVC Controllers).
```cs
public abstract class RequestHandlerBase
{
    protected RequestContext Context { get; private set; }

    protected virtual void OnContextSet();
}
```
Here, `Context` will definitely be initialized, albeit with a little bit of magic (omitted for brevity) and only *after* construction. Then, if a derived class wants to populate some of their own properties that are dependent on values in `Context`, they can override the virtual method where they *know* the context has been set, but wouldn't want to be burdened with doing explicit checks. The only caveat derived classes have to take into account is `Context` still being uninitialized while their constructor runs, but beyond that one, the property is for all intents and purposes always non-null.

**Comment by PathogenDavid** [comment]
>Sounds like an anti-pattern to require derived classes to initialize their base class properly.

It's definitely not the most kosher of things, but that doesn't mean people won't do it.

In our case, we have an abstract class `DxTexture` that represents a DirectX texture and it has all of the public functionality necessary for interacting with any texture (which isn't a lot, but it's still boilerplate.) This functionality needs access to handles relating to the texture, so `DxTexture` has two protected fields for this purpose.

From that, there's then 3 concrete implementations of this abstract class that represent 1D, 2D, and 3D textures. (The only differences are what validation is performed and how the protected fields are initialized in the constructor.)

-----

There's definitely some alternatives to this design. (In particular, having the implementations provide a factory method for creating the dimension-specific handles - which is probably what this will be changed to.) However, nullable reference types could make this type of pattern much safer by alerting implementations that they haven't provided all the necessary resources that the base implementation expects.

In fact, if this issue is fixed, I'd say nullable reference types would fix the main issue of partial initialization with this abstract-base-class-provides-boilerplate pattern.
